controller/system_rollout_controller.go

```diff
@@ -0,0 +1,1801 @@
+		}
+
+		c.recordErrorState(systemRestore, c.cacheErrors.Join(), log)
+
+	} else if c.cacheState.recordType == systemRolloutRecordTypeNormal {
+		c.recordNormalState(systemRestore, c.cacheErrors.Join(), log)
+	}
+
+	if !reflect.DeepEqual(existing.Status, systemRestore.Status) {
+		_, err = c.ds.UpdateSystemRestoreStatus(systemRestore)
+		if err != nil {
+			log.Warnf(SystemRolloutMsgEnqueueFmt, SystemRolloutErrUpdateSystemRestoreStatus)
+			c.enqueue()
+		}
+	}
+}
+
+func (c *SystemRolloutController) cacheStateUpdate(recordType systemRolloutRecordType, nextState longhorn.SystemRestoreState, reason, message string) {
+	c.cacheState.lock.Lock()
+	defer c.cacheState.lock.Unlock()
+
+	c.cacheState.recordType = recordType
+	c.cacheState.nextState = nextState
+	c.cacheState.reason = reason
+	c.cacheState.message = message
+}
+
+func (c *SystemRolloutController) cacheStateReset() {
+	c.cacheStateUpdate(systemRolloutRecordTypeNone, longhorn.SystemRestoreStateNone, "", "")
+}
+
+func (c *SystemRolloutController) getLoggerForSystemRollout() *logrus.Entry {
+	log := c.logger.WithField("systemRestore", c.systemRestore.Name)
+
+	if c.backupTargetURL != "" {
+		log = log.WithField("backupTargetURL", c.backupTargetURL)
+	}
+
+	if c.engineImage != "" {
+		log = log.WithField("engineImage", c.engineImage)
+	}
+
+	return log
+}
+
+func (c *SystemRolloutController) recordErrorState(systemRestore *longhorn.SystemRestore, err string, log logrus.FieldLogger) {
+	systemRestore.Status.State = longhorn.SystemRestoreStateError
+	systemRestore.Status.Conditions = types.SetCondition(
+		systemRestore.Status.Conditions,
+		longhorn.SystemRestoreConditionTypeError,
+		longhorn.ConditionStatusTrue,
+		c.cacheState.reason,
+		fmt.Sprintf("%v: %v", c.cacheState.message, err),
+	)
+
+	log.WithError(fmt.Errorf(err)).Error(c.cacheState.message)
+	c.eventRecorder.Eventf(systemRestore, corev1.EventTypeWarning, EventReasonFailed, util.CapitalizeFirstLetter(c.cacheState.message))
+
+	c.cacheErrors.Reset()
+}
+
+func (c *SystemRolloutController) recordNormalState(systemRestore *longhorn.SystemRestore, err string, log logrus.FieldLogger) {
+	systemRestore.Status.State = c.cacheState.nextState
+	log.Info(c.cacheState.message)
+	c.eventRecorder.Eventf(systemRestore, corev1.EventTypeNormal, c.cacheState.reason, c.cacheState.message)
+}
+
+func (c *SystemRolloutController) systemRollout() error {
+	log := c.getLoggerForSystemRollout()
+
+	systemRestore, err := c.ds.GetSystemRestore(c.systemRestoreName)
+	if err != nil {
+		return err
+	}
+
+	existingSystemRestore := systemRestore.DeepCopy()
+	defer func() {
+		c.handleState(systemRestore, existingSystemRestore, err, log)
+	}()
+
+	switch systemRestore.Status.State {
+	case longhorn.SystemRestoreStatePending:
+		systemRestore, err = c.initializeSystemRollout(systemRestore, log)
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize system rollout")
+		}
+
+		c.cacheStateUpdate(
```

Comment:
Does this controller need `cacheState`...?

Reply:
No, fixed.